### PR TITLE
[#67] Add build hook (main)

### PR DESCRIPTION
**Pull request by alanking**
Addresses #67 

Still confirming for all the platforms...
- [x] Ubuntu 18
- [x] Ubuntu 20
- [x] Ubuntu 22
- [x] Debian 11
- [x] Debian 12
- [x] CentOS 7
- [x] EL8
- [x] EL9

### CODE REVIEW

File: irods_consortium_continuous_integration_build_hook.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import glob
+import multiprocessing
+import optparse
+import os
+import tempfile
+
+import irods_python_ci_utilities
+
+def add_cmake_to_front_of_path():
+    cmake_path = '/opt/irods-externals/cmake3.21.4-0/bin'
```
Comment (alanking) [inline irods_consortium_continuous_integration_build_hook.py:12]: Need to try using a newer CMake when available...

See https://github.com/irods/irods_rule_engine_plugin_python/commit/674f5657e6d5d67c886684c32f9a9236884944c7

**Comment by korydraughn** [comment]
Does it work for each OS?

File: irods_consortium_continuous_integration_build_hook.py (hunk 1 above)
Comment (alanking) [reply irods_consortium_continuous_integration_build_hook.py:12]: Eh, nevermind. We can just piggy-back off of the effort done for other repos if this is not needed to build for the currently supported platforms.

**Comment by alanking** [comment]
Tracking progress on testing for all the platforms via checklist in OP. Building irods for all the platforms is gonna take a while, so... hang tight.

**Comment by alanking** [comment]
Everything worked except for Centos 7:
```
stderr: CMake Error at /opt/irods-externals/cmake3.21.4-0/share/cmake-3.21/Modules/FindPackageHandleStandardArgs.cmake:230 (message):   
  Could NOT find FLEX: Found unsuitable version "2.5.37", but required is at                                                            
  least "2.6.1" (found /usr/bin/flex)                
Call Stack (most recent call first):    
  /opt/irods-externals/cmake3.21.4-0/share/cmake-3.21/Modules/FindPackageHandleStandardArgs.cmake:592 (_FPHSA_FAILURE_MESSAGE)          
  /opt/irods-externals/cmake3.21.4-0/share/cmake-3.21/Modules/FindFLEX.cmake:264 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)                    
  parser/CMakeLists.txt:1 (find_package)                                                                                                
```
So, here are the options as I see it:
1. Go find the appropriate version of flex and install it in the build hook
2. Introduce flex as an external and/or build it ourselves in the project
3. Don't support Centos 7 for GenQuery2

**Comment by alanking** [comment]
Or, a fourth option... determine whether the specified minimum version is indeed the minimum required version.

**Comment by trel** [comment]
who's declaring that minimum?

**Comment by alanking** [comment]
The requirement is declared here:
https://github.com/irods/irods_api_plugin_genquery2/blob/d08182c017e88d86f6d680fde17dc346446ff88f/parser/CMakeLists.txt#L1

And using the version provided by Centos 7 causes the build to fail, so... whatever the minimum is, it's something newer than the one provided by Centos 7 :)

**Comment by alanking** [comment]
Following suggestion from @SwooshyCueb worked for Centos 7

**Comment by korydraughn** [comment]
Excellent! Looking over build hook.

**Review by korydraughn — COMMENTED** [review]
Code looks fine. If there's nothing left to do for this, squash it.

**Comment by alanking** [comment]
Squashed

**Comment by trel** [comment]
looking good.

**Review by korydraughn — APPROVED** [review]
Pound it.

**Comment by alanking** [comment]
#'d. I guess I'll merge it too :p

**Pull request merged**